executor/join.go

```diff
@@ -209,9 +264,118 @@ func (e *HashJoinExec) fetchOuterRows(goCtx goctx.Context) {
 	}
 }
 
-// prepare runs the first time when 'Next' is called, it starts one worker goroutine to fetch rows from the big table,
-// and reads all data from the small table to build a hash table, then starts multiple join worker goroutines.
-func (e *HashJoinExec) prepare(goCtx goctx.Context) error {
+// fetchOuterChunks get chunks from fetches chunks from the big table in a background goroutine
+// and sends the chunks to multiple channels which will be read by multiple join workers.
+func (e *HashJoinExec) fetchOuterChunks(goCtx goctx.Context) {
+	defer func() {
+		for i := range e.outerResultChs {
+			close(e.outerResultChs[i])
+		}
+		e.workerWaitGroup.Done()
+	}()
+	for {
+		if e.finished.Load().(bool) {
+			return
+		}
+		var outerResource *execWorkerResult
+		ok := true
+		select {
+		case <-e.closeCh:
+			return
+		case outerResource, ok = <-e.outerChkResourceCh:
+			if !ok {
+				return
+			}
+		}
+		outerResult := outerResource.chk
+		err := e.outerExec.NextChunk(goCtx, outerResult)
+		if err == nil && outerResult.NumRows() == 0 {
+			return
+		}
+		if err != nil {
+			e.joinResultCh <- &execWorkerResult{
+				err: errors.Trace(err),
+			}
+			return
+		}
+		outerResource.cycleChan <- outerResult
+	}
+}
+
+// fetchSelectedInnerRows fetches all the selected rows from inner executor,
+// and append them to e.innerResult.
+func (e *HashJoinExec) fetchSelectedInnerRows(goCtx goctx.Context) (err error) {
+	innerExecChk := e.childrenResults[e.innerIdx]
+	selected := make([]bool, 0, chunk.InitialCapacity)
+	innerResult := chunk.NewList(e.innerExec.Schema().GetTypes(), e.maxChunkSize)
+	for {
+		innerExecChk.Reset()
+		err = e.innerExec.NextChunk(goCtx, innerExecChk)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if innerExecChk.NumRows() == 0 {
+			break
+		}
+		selected, err = expression.VectorizedFilter(e.ctx, e.innerFilter, innerExecChk, selected)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		for idx, ok := range selected {
+			if ok {
+				innerResult.AppendRow(innerExecChk.GetRow(idx))
+			}
+		}
+	}
+	e.innerResult = innerResult
+	return nil
+}
+
+func (e *HashJoinExec) initializeForProbe() {
+	e.outerResultChs = make([]chan *chunk.Chunk, e.concurrency)
```

Comment:
for all the channels created in this function, maybe it's better for understanding to add comments for each of them to explain who will read them and who will write them ?